cpp/src/arrow/acero/query_context.cc

```diff
@@ -23,6 +23,36 @@ namespace arrow {
 using arrow::internal::CpuInfo;
 namespace acero {
 
+namespace internal {
+
+int64_t GetTempStackSizeFromEnvVar() {
+  auto maybe_env_value = arrow::internal::GetEnvVar(kTempStackSizeEnvVar);
+  if (!maybe_env_value.ok()) {
+    return kDefaultTempStackSize;
+  }
+  auto env_value = *std::move(maybe_env_value);
+  if (env_value.empty()) {
+    return kDefaultTempStackSize;
+  }
+
+  int64_t temp_stack_size = 0;
+  size_t length = 0;
+  bool exception = false;
+  try {
+    temp_stack_size = std::stoll(env_value.c_str(), &length);
+  } catch (const std::exception&) {
+    exception = true;
+  }
+  if (length != env_value.length() || exception || temp_stack_size <= 0) {
+    ARROW_LOG(WARNING) << "Invalid temp stack size provided in " << kTempStackSizeEnvVar
+                       << ". Using default temp stack size: " << kDefaultTempStackSize;
+    return kDefaultTempStackSize;
+  }
+  return temp_stack_size;
```

Comment:
You probably want to limit it to some safe upper-bound here

`return std::min(temp_stack_size, kMaxTempStackSize);`

Reply:
Given that: 1) a reasonable operator should use about 10s of KB - let's say 64KB; 2) a reasonable plan (not insanely complex) should contain no more than 100s of chained operators - let's give it 1024. Then a reasonable upper-bound could be 64MB?

Reply:
256GB per-thread should be enough for everyone :)

Reply:
I've updated the PR with 64MB upper-bound. Let me know if you still have concerns about this. Thanks @felipecrv !

Reply:
Should that fill better only in debug mode?

Reply:
Why is it not zero-initialized instead?

Reply:
Yes, this is for another PR.

Reply:
Issue #41460 filed for this further enhancement.